### DISCUSSION TITLE
Keep-Alive Semantics

### DISCUSSION THREAD

**Discussion by tailhook** [geldata/gel#1914] [Pre-RFC]
# Abstract

This Pre-RFC describes protocol for implementing application level keep-alive messages and provides a way to describe keep-alive timeouts.

# Motivation

While TCP provides some level of keep-alive functionality built-in. That functionality is (a) has too large timeouts by default and (b) in the presense of intermediaries setting keep-alive parameter works only to next hop. So it's common to implement end-to-end application level ping/pong messages.

# Protocol Format

The following protocol changes should be done:

1. Bump the protocol version
2. Add extension called `idle_timeout`
3. Within server response of `idle_timeout` extension set header `timeout=60` where value is stringified number of seconds of server's idle timeout
4. Add Ping/Pong messages

Message types described with protocol [conventions](https://www.edgedb.com/docs/internals/protocol/overview):
```
struct Ping {
  // Message type ('g').
  uint8           mtype = 0x67;

  // Length of message contents in bytes,
  // including self.
  uint32          message_length;

  // Data. Should be treated as opaque and send in Pong response
  byte            data[message_length];
}

struct Pong {
  // Message type ('G').
  uint8           mtype = 0x47;

  // Length of message contents in bytes,
  // including self.
  uint32          message_length;

  // Data sent from Ping. Or just empty if this is single-way Pong.
  byte            data[message_length];
}
```

The maximum size of payload (`message_length`) of Ping and Pong packets is `125` bytes.

# Protocol Semantics

There are following rules for Ping/Pong messages:

1. Ping can arrive at any time, including between sending request and receiving response (but not in the middle of the message).
2. Pong should be sent as quick as practical, but may be ignored if other message is being sent or is going to be sent at that time (i.e. common case is to discard Ping if output buffer is full).
3. Pong response must contain the contents sent in Ping.
4. Ping and Pong messages can be send from any side. Server must respond to Ping, while generally server sends Pings and client responds.
5. Server must not drop connection if Pong is missing but any other message was received (So basically Pong resets idle timeout).
6. Bare Pong (without Ping) can be sent from any side too (in this case it should contain no data).

If `idle_timeout` extension is negotiated:

1. Server do not send Ping requests.
2. Server shuts down idle connection on the timeout announced.
3. Client should either Ping and wait for Pong or assume connection is closed if it was idle for `server_timeout - max_latency` seconds.
3. Client may send Ping or Pong requests to extend idle time.
4. Server responds to Ping (and does that irregardless of if there are messages in the output buffer).

It's fine if approximation of `max_latency` is static, like 5 seconds.

By default, if `idle_timeout` extension is **not** announced, it's expected that client is able to respond to asynchronous ping requests, so they are sent by the server on the regular intervals when connection is idle for a while.

Note: idle timer is reset in two cases:
1. Any byte received from the connection
2. Connection is doing `Execute` and any chunk of data is sent (this is needed for huge queries where a lot of data is being sent, while protocol is structured in a way that nothing is received in the meantime)

# Server Shutting Down Connection

When server shuts down connection it should send [`ErrorMessage`](https://www.edgedb.com/docs/internals/protocol/messages#ref-protocol-msg-error ) with severity of `FATAL` and one of the three errors:
1. `ShutdownRequestCanceled` -- shutdown of connection which is waiting on `Execute` (or similar query executing request) and no connection is acquired from pool to make this request (i.e. server is sure that request didn't start)
2. `ShutdownRequestInterrupted` -- shutdown of connection which is in the middle of request processing (server should cancel postgres query in this case too, but we still don't know if that is applied)
3. `ShutdownClean` -- otherwise

Shutdown ErrorResponse message is sent to the connection, expecting either connection closed or `Terminate` message received in 5 seconds, otherwise connection is closed. The message is only sent if output buffer is not full. Otherwise connection is closed immediately. 

# First Implementation

1. Server sends `idle_timeout/timeout` of 60 seconds
2. Otherwise server sends `Ping` after 45 seconds of inactivity and shuts down connection if nothing is received for 15 seconds

Eventually server should send current timestamp inside the `Ping` timeout and therefore estimate latency to clients, and provide statistics about that. But that requires general interface for statistics.

# Alternatives

## Negotiating Extension

Support of ping messages could be negotiated as the protocol extension instead of bumping the protocol version and making ping support default.

Note: we don't have any other mechanism of notifying the client about idle timeout without extension.

## Sending Terminate Packet

Instead of sending ErrorResponse when closing connection server could send some kind of Terminate packet highlighting the termination reason.

## Not Sending Terminate Packet

While it's totally acceptable for server. It maybe bad idea in cases where:
1. (Sync) client sends request
2. Server is busy and don't have any connections available for 60 seconds
3. Server closes connection because connection is idle.

In this case client should be notified that their request is cancelled and may be retried cleanly.

## Not Interrupting `Execute`ing Queries

This could mean that we don't have `ShutdownRequestInterrupted` error and one less failure state to handle on client.

The problem of not doing that is the following case:

1. Request started to execute
2. Response is started being sent
3. Client hangs due to code failure
4. Response buffer is filled up and connection could wait on this state arbitrarily long, this includes reserving postgres connection for that time

**Updates:**
1. Max ping/pong packet size is set
2. Add a message on connection shut down
3. Add a clarification that connection timeout resets on the outgoing packet

**Comment by 1st1** [comment]
> while generally server sends Pings and client responds.

I'd like to see some argumentation why this should be the server's job.

**Reply by tailhook** [reply]
Motivation is:

1. We don't want to depend on client implementation for things that may leak server memory
2. We may want to tweak ping interval under heavier load or other cases, and updating configuration on every client is not viable in this case
3. We want to cut client configuration to the minimum possible (so ping interval in client config is a bad idea)

This is especially important on cloud, where one party controls EdgeDB side of things and other party (or multiple) control clients.

**Reply by 1st1** [reply]
> We don't want to depend on client implementation for things that may leak server memory

This is a very good argument for it, indeed.

**Comment by 1st1** [comment]
> Ping can arrive at any time

If we implement cancellation the reply to a pong can be lost, in cases when the client pipelines and sends

`PARSE EXECUTE' EXECUTE PING PARSE CANCEL(EXECUTE')`

IMO the idea of `Ping` is to receive *some* kind of response from the server and we should explain the caveats of sending pings along with other messages.  Ideally the client should only send pings when the connection has been idle for some time and no side is actively engaged with another.

**Reply by tailhook** [reply]
> If we implement cancellation the reply to a pong can be lost

I'm not sure why. Are we supposed to clean network buffers when cancel is received? I don't think this is a viable solution as we can clean partial packets. If we don't clear network buffers there's nothing that stops from sending Pong I think.

>  Ideally the client should only send pings when the connection has been idle for some time and no side is actively engaged with another.

This is hard to specify and next to impossible to enforce. Especially if we ever implement notifications in edgedb or if next data packet delayed in postgres for too long due to massive disk usage.

**Reply by 1st1** [reply]
> I'm not sure why. Are we supposed to clean network buffers when cancel is received? I don't think this is a viable solution as we can clean partial packets. If we don't clear network buffers there's nothing that stops from sending Pong I think.

No, we wouldn't clean the buffer, we'd just ignore most of the messages. That said, we can make a special rule for pings. I think isn't a problem after all.

> This is hard to specify and next to impossible to enforce. Especially if we ever implement notifications in edgedb or if next data packet delayed in postgres for too long due to massive disk usage.

*nod*

**Comment by 1st1** [comment]
> Pong response must contain the contents sent in Ping.

I think we should put a limit on how big Ping messages can get.

**Reply by tailhook** [reply]
~~I'm not sure. WebSockets don't have any limit on Ping/Pong packets apart from saying that control frames cannot be fragmented (and fragments can be 2^64 size, except some implementations can impose limits on a single fragment).~~


~~I'm not sure any arbitrary limit will do any good for us. What are we protecting from? There are better ways to DoS connection, including sending lots of Ping packets if there is a limit imposed on single one. Or does this simplify managing buffering somewhere?~~


**Update:** Websockets do have limit of 125 bytes on control frames. I'm adding the same limit, it's kinda enough and is useful to make 1:1 mapping with WebSockets if we ever want that.

**Reply by 1st1** [reply]
Great!

**Comment by 1st1** [comment]
> Bare Pong (without Ping) can be sent from any side too (in this case it should contain no data).

What's the motivation for this?

**Reply by tailhook** [reply]
This goes straight from WebSocket spec. The idea is that if your peer has connection closed but did not send TCP Reset message for some reason (hardware failed and restarted; new VM on that IP; or just packet was dropped few times in row), sending any data like Pong packet will trigger TCP Reset from the peer and close the connection.

Use cases:
1. Sync clients that do not want to wait for the whole Ping/Pong roundtrip to promote idle timeout
2. Sync and async clients that don't want to bother with Pings and Pongs in the middle other request processing can just periodically send Pongs (although I would not recommend this)

**Reply by 1st1** [reply]
> sending any data like Pong packet will trigger TCP Reset from the peer and close the connection.

Interesting. But instead of sending Pong we could just send Ping then?

**Reply by tailhook** [reply]
Yes, but then we have either:
1. wait for Ping right away, or
2. expect Pong when sending next Prepare (or any other) message, as we didn't parse Pong after Ping. So any code that waits for specific response need to skip Pong messages.

**Comment by 1st1** [comment]
I'm having trouble following through

```
If idle_timeout is announced:

    Server do not send Ping requests.
```

What does it mean to announce `idle_timeout`? Enable the `idle_timeout` protocol extension?  Re "Server do not send Ping requests." -- just a few sentences before it was said that both Client and Server can send Pings, no?

**Reply by tailhook** [reply]
Changed to "If idle_timeout extension is negotiated:". 

> just a few sentences before it was said that both Client and Server can send Pings, no?
 
Yes. By the protocol spec both can send Pings. But `idle_timeout` extension disables actually sending them by server to make implementing sync clients a little bit easier.

**Reply by 1st1** [reply]
Oh, I totally forgot about sync clients, where responding to a Ping can only happen if

* there's a need for the client to send a request (and it discovers there's a ping message from the server in the buffer)
* the client uses an OS thread to listen to the socket (relatively unlikely we'll bother with this).

So essentially you're saying that the server is free to disconnect the client if it doesn't hear from it for 60 seconds, which would mean even for the sync connections that they're not actively used. Right?

**Reply by tailhook** [reply]
Exactly.

The point of showing timeout value to the client is that client don't have race condition where it started sending request and connection was closed on the other side (in this case case there is no way to find out whether request worked on server or not). And regulate that timeout from server even for sync clients.

**Comment by elprans** [comment]
>     3. extension set header `timeout=60` where value is stringified number of seconds

I think there was some discussion on making headers properly typed (i.e. use the normal data encodings).

>     1. Server sends `idle_timeout/timeout` of 60 seconds

IMO we should make all timeouts/periods configurable from the get go.

**Reply by tailhook** [reply]
> I think there was some discussion on making headers properly typed

Yes. But until we decided how that works, I don't think we should create a mix of ways how headers are typed. We already have string headers, so at least we don't introduce new encodings.

Also property typed headers are more important where structure is more complex (i.e. list of implicit fields in query, or readonly metrica settings).

> IMO we should make all timeouts/periods configurable from the get go.

It's possible but there are downsides:
1. People may start configuring these values themselves and start relying on configured timeouts, which are possibly wrong
2. Users may want setting these settings on the cloud, but we want to control timeouts on the cloud to keep resource usage at optimal level
3. Changing defaults in later versions would make more surprizes 

Generally, I don't see why timeouts are infrastructure- or application-dependent. We can't set subsecond timeouts, because of how TCP works (in particular retransmission timeout). We can't set it too high because that make them useless. So the range of valid values is not too large.

Generally, applications should work fine with any sane timeouts. The only fragile case is idle timeout breaking open transaction. And this is important for both sides:
1. We should not make idle timeout too small
2. And users should not set timeouts that "work on their laptop"

**Comment by tailhook** [comment]
I've updated the spec to include `ErrorResponse` in connection shutdown message, because that is what is written in spec:

>In some cases the server might disconnect without a client request to do so. In such cases the server will attempt to send an ErrorResponse or a LogMessage message to indicate the reason for the disconnection.


But I'm not sure this is a good idea. In some sense it would be better to have specialized `Terminate` message (we only have client -> server one now.

**Comment by 1st1** [comment]
> ShutdownRequestCanceled -- shutdown of connection which is waiting on Execute (or similar query executing request) and no connection is acquired from pool to make this request (i.e. server is sure that request didn't start)

~I'm not sure I understand the "i.e. server is sure that request didn't start" part.~

NVM, I got it.

**Comment by 1st1** [comment]
@elprans and I have brainstormed this RFC today, here's the summary.

One of the key reasons for this proposal is to have a way for the server to terminate no longer active connections. The proposed solution is for the server to send periodic PING messages and terminate the connection if the client fails to respond in time. Two problems here:

1. There's no way for blocking IO clients to respond to server PINGs. Therefore the plan is to simply close them after some period of inactivity. It's hard to estimate what proportion of EdgeDB clients will be non-blocking, but it's worth mentioning that the async/await trend is still relatively new.

2. The fact that a non-blocking connection responds to server PINGs doesn't always mean much. The connection can be accidentally left open and kept alive by the GC, and most async frameworks would allow it to happily respond to server PINGs, even if it's no longer in active use.

With (1) and (2) in mind it sounds that the easiest and most effective solution is to:

* Add a new config setting: `idle_timeout`. We can set it to 60 seconds by default. Setting it to 0 would disable the mechanism. If a client hasn't sent any messages within this time window the server can terminate the connection. The config setting will be announced to the client at the connection time.

* We allow clients to send PING messages. This would allow applications like REPL, which might have complex connection state, to maintain their connections open as long as they need, by periodically sending PINGs to the server.

* There will be no PING messages sent by the server. There seems to be no real need for them -- inactivity timeouts achieve essentially the same goal.

**Reply by elprans** [reply]
Quick add to further clarify why `idle_timeout` + client-sent `Ping` are a good solution.

There are two competing goals here:

1. The server wants to conserve resources and terminate dangling connections ASAP.  From its point of view it's the _client's_ job to maintain an active connection.  And so, a simple `idle_timeout` is the simplest and most reliable way to terminate inactive sessions.

2. (Some) clients don't want to be affected by a session termination due to inactivity.  The design discussed in https://github.com/edgedb/edgedb/discussions/1896 will solve this naturally for most applications that use language bindings: the client will simply automatically reconnect on the next query.  For applications such as the REPL, where long periods of inactivity are expected and there might also be non-trivial session state set up by the user, there is an explicit remedy to reset server idle timeout: a `Ping` message (we also considered sending dummy queries, but that seems like an unnecessary (however small) load on the database backend).

**Reply by tailhook** [reply]
While I understand your solution. There are couple of issues, though:
1. We rely on client implementation to correctly estimate roundtrip interval and send ping before idle timeout happens
2. There is the risk, that stackoverflow response for "why connections are dropped?" and "why edgedb is slow after a while?" will be disable idle_timeout. Which is very wrong. Actually I don't see when disabling idle timeout makes any sense (i.e. when TCP keep-alives are set up and that interval is fine, and there are no intermediaries?).
3. On cloud we don't have statistics on:
    a) how many async clients there are
    b) what is average roundtrip time to the client
4. So effectively, on cloud: we can't fine tune pings and idle timeouts, and rely on users to configure them correctly (i.e. for edgedb instance that is shared between users this can be a problem).
5. When server sends pings we can employ the following strategy on connection pressure:
    a) let's say we have 700 out of limit of 1000 connections open
    b) another 300 goes into quick succession, filling all the input slots
    c) server pings older 700 connections ahead of time
    d) all connections that didn't respond on the ping in (c) are closed
    This could help recover from some network splits much faster. While admittedly this is mostly a speculation, this just shows that more interesting strategies could be tried with server-initialed pings.

I don't think that leaked client waiting for GC is a problem in my solution any more than in yours:
1. We're likely to enable pings by default
2. Such client will equally well send pings as respond to pongs

Also pings from server to client should not protect from hanging clients. Client that don't do request are quite cheap. What it has to protect from is hundreds of VMs starting and crashing. Or network issues that cause connection reset packets to drop.

So generally your proposal is a subset of mine. And I don't see the downsides of mine.

**Reply by 1st1** [reply]
> There is the risk, that stackoverflow response for "why connections are dropped?" and "why edgedb is slow after a while?" will be disable idle_timeout. Which is very wrong. 

Well, the idea with your other proposals is to make our APIs somewhat resilient to raw connection drops.

> Actually I don't see when disabling idle timeout makes any sense (i.e. when TCP keep-alives are set up and that interval is fine, and there are no intermediaries?).

Me neither, but some might want to tweak the default, so we'll allow that. We can make it impossible to disable it though, meaning that users will just have to specify some ridiculously long timeouts if they don't want the mechanism -- e.g. set the timeout to 7 days which would effectively disable it.

> So effectively, on cloud: we can't fine tune pings and idle timeouts, and rely on users to configure them correctly (i.e. for edgedb instance that is shared between users this can be a problem).

The idea is that pings shouldn't be used by clients. Except maybe by special applications like REPLs. Non-blocking clients will not send them by default.

> I don't think that leaked client waiting for GC is a problem in my solution any more than in yours:
>
>    We're likely to enable pings by default

I don't see a use case for that. If you're idle for 60 seconds it means that most likely you're not using the connection. If disconnecting and reconnecting the underlying raw connection is made relatively transparent for the user we won't have a problem here.

> So generally your proposal is a subset of mine. And I don't see the downsides of mine.

First, both proposals are effectively yours :)  I've just tweaked it a little.

Second, the initial proposal with server-side pings totally hand-waves the fact that blocking IO still dominates non-blocking IO. And even if the split was 50/50, it's still not clear why server pings would only work for half of clients and would be totally ignored by the other half.

Both variations of the proposal solve the problem of unused resources and let the server to recycle connections relatively quickly. Not implementing server-side pings simplifies both the server implementation and implementations of clients, without sacrificing a lot of functionality.
